plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.ts

```diff
@@ -312,44 +323,120 @@ export class DefaultProcessingDatabase implements ProcessingDatabase {
-      await tx<DbRefreshStateRow>('refresh_state')
-        .insert(row)
-        .onConflict('entity_ref')
-        .merge(['unprocessed_entity', 'last_discovery_at']);
+    for (const { entity, locationKey } of options.entities) {
+      const entityRef = stringifyEntityRef(entity);
+      const serializedEntity = JSON.stringify(entity);
+
+      // We optimistically try to update any existing refresh state first, as this is by far
+      // the most common case.
+      const refreshResult = await tx<DbRefreshStateRow>('refresh_state')
+        .update({
+          unprocessed_entity: serializedEntity,
+          location_key: locationKey,
+          last_discovery_at: tx.fn.now(),
+        })
+        .where('entity_ref', entityRef)
+        .andWhere(inner => {
+          if (!locationKey) {
+            return inner.whereNull('location_key');
+          }
+          return inner
+            .where('location_key', locationKey)
+            .orWhereNull('location_key');
+        });
+
+      if (refreshResult === 0) {
+        // In the event that we can't update an existing refresh state, we first try to insert a new row
+        try {
+          let query = tx('refresh_state').insert<any>({
+            entity_id: uuid(),
+            entity_ref: entityRef,
+            unprocessed_entity: serializedEntity,
+            errors: '',
+            location_key: locationKey,
+            next_update_at: tx.fn.now(),
+            last_discovery_at: tx.fn.now(),
+          });
+
+          // TODO(Rugvip): only tested towards Postgres and SQLite
+          // We have to do this because the only way to detect if there was a conflict with
+          // SQLite is to catch the error, while Postgres needs to ignore the conflict to not
+          // break the ongoing transaction.
+          if (tx.client.config.client !== 'sqlite3') {
+            query = query.onConflict('entity_ref').ignore();
+          }
+
+          const result = await query;
+          if (result.rowCount === 0) {
+            throw new ConflictError(
+              'Insert failed due to conflicting entity_ref',
+            );
+          }
+        } catch (error) {
+          if (
+            !error.message.includes('UNIQUE constraint failed') &&
+            error.name !== 'ConflictError'
+          ) {
+            throw error;
+          }
+          // If the row can't be inserted, we have a conflict, but it could be either
+          // because of a conflicting locationKey or a race with another instance, so check
+          // whether the conflicting entity has the same entityRef but a different locationKey
+          const [conflictingEntity] = await tx<DbRefreshStateRow>(
+            'refresh_state',
+          )
+            .where({ entity_ref: entityRef })
+            .select();
+
+          // If the location key matches it means we just had a race trigger, which we can safely ignore
+          if (conflictingEntity.location_key !== locationKey) {
```

Comment:
highly unlikely, but let's make sure to handle `conflictingEntity` being undefined here

Reply:
Yep, meh at typescripts handling of arrays every now and then :<